### PR TITLE
issue-2270: Adding support for thrift

### PR DESCRIPTION
**Pull request by ananthdurai**
This PR is adding support for Thrift File format, similar to json, csv & Avro. The `ThriftRecordReader` requires fully qualified thrift generated java class name to deserialize byte stream. `ThriftRecordReaderConfig` holds the config parameters for `ThriftRecordReader`.

### CODE REVIEW

File: pinot-core/src/main/java/com/linkedin/pinot/core/data/readers/ThriftRecordReader.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+            if (_fieldNameToIndexMap.containsKey(fieldName)) {
+                int tFieldId = _fieldNameToIndexMap.get(fieldName);
+                TFieldIdEnum tFieldIdEnum = t.fieldForId(tFieldId);
+                Object thriftValue = t.getFieldValue(tFieldIdEnum);
+                Object value;
+                if (fieldSpec.isSingleValueField()) {
+                    String token = thriftValue != null ? thriftValue.toString() : null;
+                    value = RecordReaderUtils.convertToDataType(token, fieldSpec);
+                } else {
+                    value = RecordReaderUtils.convertToDataTypeArray((ArrayList) thriftValue, fieldSpec);
```
Comment (kishoreg) [inline pinot-core/src/main/java/com/linkedin/pinot/core/data/readers/ThriftRecordReader.java]: will thriftValue always be an ArrayList?

File: pinot-core/src/main/java/com/linkedin/pinot/core/data/readers/ThriftRecordReader.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    public Schema getSchema() {
+        return _schema;
+    }
+
+    @Override
+    public void close() throws IOException {
+        _bufferIn.close();
+    }
+
+    private Class<TBase<?, ?>> thriftClassCreator() {
```
Comment (kishoreg) [inline pinot-core/src/main/java/com/linkedin/pinot/core/data/readers/ThriftRecordReader.java]: rename to initThriftInstanceCreator

File: pinot-core/src/main/java/com/linkedin/pinot/core/data/readers/ThriftRecordReaderConfig.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+package com.linkedin.pinot.core.data.readers;
+
+import org.apache.commons.lang3.builder.ToStringBuilder;
+import org.apache.commons.lang3.builder.ToStringStyle;
+
+public class ThriftRecordReaderConfig implements RecordReaderConfig {
+
+    private String _thriftClass;
+
+    public String get_thriftClass() {
```
Comment (kishoreg) [inline pinot-core/src/main/java/com/linkedin/pinot/core/data/readers/ThriftRecordReaderConfig.java]: getThriftClass

File: pinot-core/src/main/java/com/linkedin/pinot/core/data/readers/ThriftRecordReaderConfig.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        return _thriftClass;
+    }
+
+    public void set_thriftClass(String _thriftClass) {
```
Comment (kishoreg) [inline pinot-core/src/main/java/com/linkedin/pinot/core/data/readers/ThriftRecordReaderConfig.java]: setThriftClass also change the method param name  to thriftClass

File: pinot-core/src/test/java/com/linkedin/pinot/core/data/readers/ThriftRecordReaderTest.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Test {@code com.linkedin.pinot.core.data.readers.ThriftRecordReader} for a given sample thrift
+ * data Sample code to generate thrift test data.
+ *
+ * <pre>
+ * {@code
+ * ThriftSampleData t1 = new ThriftSampleData();
```
Comment (kishoreg) [inline pinot-core/src/test/java/com/linkedin/pinot/core/data/readers/ThriftRecordReaderTest.java]: we can probably generate the file as part of the test case beforeTest method

File: pinot-core/src/main/java/com/linkedin/pinot/core/data/readers/RecordReaderUtils.java [hunk 6]
```diff
@@ -33,13 +35,21 @@ private RecordReaderUtils() {
...
       inputStream = new GZIPInputStream(new FileInputStream(dataFile));
     } else {
       inputStream = new FileInputStream(dataFile);
     }
-    return new BufferedReader(new InputStreamReader(inputStream, "UTF-8"));
+    return inputStream;
+  }
+
+  public static BufferedInputStream getFileBufferStream(File dataFile) throws IOException {
+    return new BufferedInputStream(getFileStreamReader(dataFile), 2048);
```
[the remark this answers is not in the document]
Comment (ananthdurai) [reply pinot-core/src/main/java/com/linkedin/pinot/core/data/readers/RecordReaderUtils.java]: My bad, I changed during one of the debug. Will revert it.

File: pinot-core/src/main/java/com/linkedin/pinot/core/data/readers/ThriftRecordReader.java (hunk 1 above)
Comment (ananthdurai) [reply pinot-core/src/main/java/com/linkedin/pinot/core/data/readers/ThriftRecordReader.java]: Thrift support `ArrayList`, `HashSet` and `HashMap`. I will add support for `HashSet` and exclude `HashMap` for now.

**Pull request merged**